src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
+                        {
+                            isCompleted = true;
+                            return TaskCache.CompletedTask;
+                        }
+
+                        if (exceptionContext != null)
+                        {
+                            if (exceptionContext.Result != null)
+                            {
+                                goto case State.ExceptionShortCircuit;
+                            }
+
+                            Rethrow(exceptionContext);
+                        }
+
+                        goto case State.ResultBegin;
+                    }
+
+                case State.ActionBegin:
+                    {
+                        var controllerContext = _controllerContext;
+
+                        _cursor.Reset();
+
+                        _controller = _controllerFactory.CreateController(controllerContext);
+
+                        _arguments = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+                        var task = _controllerArgumentBinder.BindArgumentsAsync(controllerContext, _controller, _arguments);
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ActionNext;
+                            return task;
+                        }
+
+                        goto case State.ActionNext;
+                    }
+
+                case State.ActionNext:
+                    {
+                        var current = _cursor.GetNextFilter<IActionFilter, IAsyncActionFilter>();
+                        if (current.FilterAsync != null)
+                        {
+                            if (_actionExecutingContext == null)
+                            {
+                                _actionExecutingContext = new ActionExecutingContext(_controllerContext, _filters, _arguments, _controller);
+                            }
+
+                            state = current.FilterAsync;
+                            goto case State.ActionAsyncBegin;
+                        }
+                        else if (current.Filter != null)
+                        {
+                            if (_actionExecutingContext == null)
+                            {
+                                _actionExecutingContext = new ActionExecutingContext(_controllerContext, _filters, _arguments, _controller);
+                            }
+
+                            state = current.Filter;
+                            goto case State.ActionSyncBegin;
+                        }
+                        else
+                        {
+                            goto case State.ActionInside;
+                        }
+                    }
+
+                case State.ActionAsyncBegin:
+                    {
+                        Debug.Assert(_actionExecutingContext != null);
+
+                        var filter = (IAsyncActionFilter)state;
```

Comment:
Do we want to also do Debug.Assert(state != null or state = _cursor.CurrentFilter.AsyncFilter) (applicable to other filter stages). The goal is to ensure that we always set the state to the filter in State.ActionNext